sw/host/opentitanlib/src/bootstrap/legacy.rs

```diff
@@ -202,40 +206,74 @@ impl UpdateProtocol for Legacy {
+        // All frames up to but not including this index have been ack'ed by the bootloader.
+        // Once this reaches frames.len(), the operation was successful.
+        let mut first_unacked_index = 0;
+
+        // Counts the number of repeated failures in getting a frame ack.
+        let mut consecutive_errors = 0;
+
+        // Set if the past transaction ack'ed the block sent before that one, while transmitting
+        // the next.  If so, we heuristically assume that the block just sent will be ack'ed in
+        // the upcoming bidirectional transaction.
+        let mut optimistic = false;
+        
+        // Set if the past transaction ack'ed the block seet before that one, while
+        // re-transmitting the same block.  If so, we do not care about the ack in the upcoming
+        // bidirectional transaction.
+        let mut becoming_optimistic = false;
+
+        loop {
+            if consecutive_errors > 100 {
+                return Err(LegacyBootstrapError::RepeatedErrors.into())
+            }
+
+            let second_unacked_index = (first_unacked_index + 1).min(frames.len() - 1);
+            let transmit_index =
+                if optimistic { second_unacked_index } else { first_unacked_index };
+            let frame = &frames[transmit_index];
+            eprint!("{}.", transmit_index);
+            std::thread::sleep(self.inter_frame_delay);
+            
+            // Write the frame and read back the ack of a previously transmitted frame.
+            let mut response = [0u8; std::mem::size_of::<Frame>()];
+            spi.run_transaction(&mut [Transfer::Both(frame.as_bytes(), &mut response)])?;
+
+            if becoming_optimistic {
+                optimistic = true;
+                becoming_optimistic = false;
+                continue;
+            }
+
+            if response.iter().all(|&x| x == response[0]) {
+                // A response consisteing of all identical bytes is a status code.
+                match LegacyBootstrapError::from(response[0]) {
+                    LegacyBootstrapError::NotReady => {
+                        consecutive_errors += 1;
+                        continue; // Retry sending same frame.
                     }
+                    error => return Err(error.into()),
                 }
+            }
 
-                // Compare response with checksum of last block sent.
-                if response[..Frame::HASH_LEN] == want_hash {
-                    // Successful response, move on to the next frame.
-                    break;
-                }
-                log::error!(
-                    "Frame hash mismatch device:{:x?} != frame:{:x?}.",
-                    &response[..Frame::HASH_LEN],
-                    want_hash
-                );
-                // Retry sending same frame.
+            if response[..Frame::HASH_LEN] == frames[first_unacked_index].frame_hash() {
+                first_unacked_index = first_unacked_index + 1;
```

Comment:
This can be `+= 1;`.

Reply:
I considered that, but on the other hand, I like the symmetry of having both if statements follow the same pattern:
```
if ... frames[X]... {
  ... = X + 1;
} ...
```